ndt7/server/upload/upload.go

```diff
@@ -0,0 +1,221 @@
+	logging.Logger.Warn(message)
+	writer.Header().Set("Connection", "Close")
+	writer.WriteHeader(http.StatusBadRequest)
+}
+
+// getConnFileAndPossiblyEnableBBR returns the connection to be used to
+// gather low level stats and possibly enables BBR. It returns a file to
+// use to gather BBR/TCP_INFO stats on success, an error on failure.
+func getConnFileAndPossiblyEnableBBR(conn *websocket.Conn) (*os.File, error) {
+	fp := fdcache.GetAndForgetFile(conn.UnderlyingConn())
+	// Implementation note: in theory fp SHOULD always be non-nil because
+	// now we always register the fp bound to a net.TCPConn. However, in
+	// some weird cases it MAY happen that the cache pruning mechanism will
+	// remove the fp BEFORE we can steal it. In case we cannot get a file
+	// we just abort the test, as this should not happen (TM).
+	if fp == nil {
+		err := errors.New("cannot get file bound to websocket conn")
+		logging.Logger.WithError(err).Warn("Cannot enable BBR")
+		return nil, err
+	}
+	err := bbr.Enable(fp)
+	if err != nil {
+		logging.Logger.WithError(err).Warn("Cannot enable BBR")
+		// FALLTHROUGH
+	}
+	return fp, nil
+}
+
+// gatherAndSaveTCPInfoAndBBRInfo gathers TCP info and BBR measurements from
+// |sockfp| and stores them into the |measurement| object as well as into the
+// |resultfp| file. Returns an error on failure and nil in case of success.
+func gatherAndSaveTCPInfoAndBBRInfo(measurement *model.Measurement, sockfp *os.File, resultfp *results.File) error {
+	bbrinfo, err := bbr.GetMaxBandwidthAndMinRTT(sockfp)
+	if err == nil {
+		measurement.BBRInfo = &bbrinfo
+	}
+
+	metrics, err := tcpinfox.GetTCPInfo(sockfp)
+	if err == nil {
+		measurement.TCPInfo = &metrics
+	}
+
+	if err := resultfp.WriteMeasurement(*measurement, "server"); err != nil {
+		logging.Logger.WithError(err).Warn("Cannot save measurement on disk")
+		return err
+	}
+	return nil
+}
+
+// measuringLoop is the loop that runs the measurements in a goroutine.
+// This function exits when
+//     (1) a fatal error occurs or
+//     (2) the maximum elapsed time for the upload test expires.
+// Since this function has access to BBR stats (if available), it's the
+// right place to stop the test early. The rest of the upload code is supposed
+// to stop receiving when this function signals that we're done by closing
+// the channel.
+func measuringLoop(ctx context.Context, request *http.Request,
+	conn *websocket.Conn, dataDir string, dst chan model.Measurement) {
+	logging.Logger.Debug("Starting measurement loop")
+	defer logging.Logger.Debug("Stopping measurement loop")
+	defer close(dst)
+
+	resultfp, err := results.OpenFor(request, conn, dataDir, "upload")
+	if err != nil {
+		return // error printed already
+	}
+
+	defer iox.ErrorLoggingCloser(resultfp).Close()
+
+	sockfp, err := getConnFileAndPossiblyEnableBBR(conn)
+	if err != nil {
+		return // error printed already
+	}
+
+	defer iox.ErrorLoggingCloser(sockfp).Close()
```

Comment:
As an additional comment to this API, I find the name not super clear. I tend to think about logging stuff as less relevant and I may inadvertently zap log related stuff, while I usually religiously keep the `defer foo.Close()` bits.

Reply:
I have some additional thoughts on this `iox.ErrorLoggingCloser` API. I really don't like it (sorry). I think it makes the code more obscure and gives us a false sense of security. IMHO the code review guidlines that we'll adopt _inter nos_ should be such that we'll always ask the question whether ignoring the error of a `defer foo.Close()` is sensible or not.

When ignoring the error is sensible, we should not care much about logging. When we should not ignore the error because it indicates a specific error condition that we don't want to ignore, we should write the code such that we handle the error in some way.

Take the specific case of the disk full that I mentioned above as an example. If we cannot write results on disk, this is a serious error that we want to address, IMHO shutting down the server. Logging is not going to help in this case since we are going to be writing on a full disk anyway.

Take instead the case where we get an error when closing the connection. Do we care about writing it into the logs? I am not sure. In such case, it would probably be nicer (but way more convolute) to record in the measurement that we could not successfully close the connection. There are middleboxes around. They may mess with our TLS. So, if we want to diagnose weird cases by looking into the data, the right place where to write such information is probably into the results and not into a log file that is only local to the machine.

Basically, my argument is that you want the information to be actionable or you don't want the information. I think the logs are not the right place where to put most information. To me, logs are useful when you're developing but I do not see them much useful at scale.

(I reckon I ended up having some soapbox against this API. So, I tried to spend quite some time to explain my reasoning and I am very curious to see what others think and their counter-arguments.)

Reply:
You are half right: logs are, usually, a garbage place to put information.

However, in the new world with Kubernetes our logs are on track to becoming searchable on a worldwide basis.

I like the `ErrorLoggingCloser` in principle, but I have been thinking more about it and I maybe kind of want to change its api to make the line be:

```go
	defer iox.ErrorLoggingCloser(sockfp, "Warning: ignored error").Close()
 ```

Then we can grep our logs to see whether ignored errors actually occur.  We could also (maybe) add prometheus metrics to the ErrorLoggingCloser that could make the problems available within prometheus.  (If this were a google-internal system, we would make this code generate logs at the "WARNING" level and also have the equivalent of a prometheus CounterVec counting the ignored close errors for each context. That intuitively feels like it might be overkill here, but maybe not?)

Basically, you are right that in the past logs were dumb. We are building a system where logs are not so dumb. Your (and my) old instinct that logs aren't useful in production is soon going to become obsolete. We should no longer feel like logging things to stderr/stdout is equivalent to throwing them in the trash.

Reply:
Thanks, @pboothe. I think the availability of searchable logs is a good argument that I did not consider when thinking about the problem. This clearly makes the log actionable, which is what concerned me regarding the usage.

In this vein, yeah, a specific error message can probably be helpful. Since `apex/log` emits JSON logs as configured in `ndt7`, it may perhaps make sense to emit a warning using `apex/log` also in this case.

WRT Prometheus, I'm glad to see I guessed half the right thing to do. With searchable logs, this is probably less important in general. (I think in the specific case, we probably want to know when we cannot write to disk, and maybe an alert based on that is more graceful than `panic`.)

Speaking of the API itself, I think as a programmer not super familiar with the code base, I would probably be less surprised by this alternative API:

```Go
    defer iox.CloseAndWarnOnFailure(sockfp, "ignoring sockfp.close error")
```

because it seems to me that conveys more clearly the programmer intent. I guess that, if I see a function named like this, I would not bother with searching for the definition of it, because it is self documenting. OTOH, `iox.ErrorLoggingCloser` chained with `.Close` looks a bit more obscure to me because I need to figure out what is the return value, and I would probably need to go and read the documentation when reading the code.

Reply:
I think that third party API users are the ultimate authority on what the most readable solution is. So if you think that `CloseAndWarnOnFailure` is clearer, then so do I, as you are the closest thing we currently have to a user of the library.

Here is one last idea to stir the pot:  What if we called the library `warnonfailure` instead of `iox`? Then the code would be:

```go
defer warnonfailure.Close(sockfp, "ignoring sockfp.Close error")
```

It also suggests that other methods could be added to the library, which would more closely mirror Roberto's original intent.

Reply:
Yes, I think changing the name as you propose is great and makes the code super clear and explicit! Thanks!

Reply:
Resolving since @evfirerob implemented this suggestion.